examples/experimental/native/flexible_primitives_mnist_pytorch/model_def.py

```diff
@@ -0,0 +1,172 @@
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(256, 1),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, img):
+        img_flat = img.view(img.size(0), -1)
+        validity = self.model(img_flat)
+
+        return validity
+
+
+class MNistTrial(PyTorchTrial):
+    def __init__(self, trial_context: PyTorchTrialContext) -> None:
+        self.context = trial_context
+        self.logger = TorchWriter()
+
+        # Create a unique download directory for each rank so they don't overwrite each other.
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.data_downloaded = False
+
+        mnist_shape = (1, 28, 28)
+        self.generator =  self.context.Model(Generator(latent_dim=self.context.get_hparam("latent_dim"), img_shape=mnist_shape))
+        self.discriminator =  self.context.Model(Discriminator(img_shape=mnist_shape))
+
+        lr = self.context.get_hparam("lr")
+        b1 = self.context.get_hparam("b1")
+        b2 = self.context.get_hparam("b2")
+
+        self.opt_g = self.context.Optimizer(torch.optim.Adam(self.generator.parameters(), lr=lr, betas=(b1, b2)))
+        self.opt_d = self.context.Optimizer(torch.optim.Adam(self.discriminator.parameters(), lr=lr, betas=(b1, b2)))
+
+        self.lr_g = self.context.LRScheduler(
+            lr_scheduler=LambdaLR(self.opt_g, lr_lambda=lambda epoch: 0.95 ** epoch),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+    def build_training_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        train_data = data.get_dataset(self.download_directory, train=True)
+        return DataLoader(train_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def build_validation_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        validation_data = data.get_dataset(self.download_directory, train=False)
+        return DataLoader(validation_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def train_batch(
+        self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        imgs, _ = batch
+
+        self.generator.requires_grad_(True)
+        self.discriminator.requires_grad_(False)
+
+        # sample noise and match gpu device (or keep as cpu)
+        z = torch.randn(imgs.shape[0], self.context.get_hparam("latent_dim"))
+        z = self.context.to_device(z)
```

Comment:
This seems a little odd to me since they don't have to put the device. They need to know the device to put on? Would .cuda() work in this case - cause that would be okay if they can do that.

Reply:
Yes. The reason why they don't need to / can't put a device is that we assign a device to each process and it has to use that. `.cuda(device)` essentially is equivalent to `.to(device)`, which shouldn't be called in this function. The documentation should include that.

Reply:
Also FYI: https://discuss.pytorch.org/t/what-is-the-difference-between-doing-net-cuda-vs-net-to-device/69278
`to` is more flexible and `cuda` is more of a old version interface.

Reply:
so the user needs to put: 
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
 basically everytime?

Reply:
I mean users need to put `z = self.context.to_device(z)` every time. We handle devices for users.

Reply:
OMG I completely read this line wrong. This is fine. :)